File: mobile/app/src/main/java/vn/lailua/app/logic/WhatIf.kt
package vn.lailua.app.logic

import vn.lailua.app.data.Consequence
import vn.lailua.app.data.JunctionScene
import vn.lailua.app.data.JunctionVehicle
import vn.lailua.app.data.Question
import java.text.Normalizer

/**
 * Mô phỏng "nếu làm vậy thì sao" — chuyển từ src/lib/whatif.ts: từ đáp án người học chọn, suy ra
 * kịch bản chạy sa hình (xe nào đi trước, xe nào bị cắt ngang, xe nào vượt đèn đỏ…) để
 * JunctionCanvas diễn lại hậu quả thay vì chỉ chiếu cách đi đúng.
 */
object WhatIf {
    data class Conflict(val offender: String, val victim: String, val reason: String)

    data class Plan(
        val correct: Boolean,
        val order: List<List<String>>,
        val steps: List<String>,
        val conflict: Conflict? = null,
        val violators: List<String>,
        val verdict: Consequence,
    )

    /** Bỏ dấu tiếng Việt + chữ thường để so khớp nhãn xe trong đáp án. */
    fun fold(s: String): String =
        Normalizer.normalize(s, Normalizer.Form.NFD).replace(Regex("[\\u0300-\\u036f]"), "").replace('đ', 'd').replace('Đ', 'D').lowercase()

    private fun lowerFirst(s: String) = if (s.isEmpty()) s else s[0].lowercaseChar() + s.substring(1)

    private data class Hit(val id: String, val at: Int, val text: String)

    /** Vị trí xuất hiện của từng xe trong đoạn văn (theo nhãn), tránh nhãn lồng nhau ("Xe con" trong "Xe con A"). */
    private fun findVehicles(text: String, vehicles: List<JunctionVehicle>): List<Hit> {
        val hay = fold(text)
        val labels = vehicles.map { it.id to fold(it.label) }.sortedByDescending { it.second.length }
        val taken = ArrayList<IntRange>()
        val hits = ArrayList<Hit>()
        for ((id, key) in labels) {
            if (key.isEmpty()) continue
            var from = 0
            while (from < hay.length) {
                val i = hay.indexOf(key, from)
                if (i < 0) break
                val overlaps = taken.any { i < it.last + 1 && i + key.length > it.first }
                if (!overlaps) {
                    taken += i until i + key.length
                    hits += Hit(id, i, hay)
                    break
                }
                from = i + 1
            }
        }
        return hits.sortedBy { it.at }
    }

    data class Parsed(val groups: List<List<String>>, val movesAll: Boolean, val stopsAll: Boolean)

    /** Phân tích một đáp án thành các nhóm xe đi lần lượt; xe trong phần "…; xe X dừng lại" không đi. */
    fun parseOption(option: String, spec: JunctionScene): Parsed? {
        val parts = option.split(";")
        val main = parts.first()
        val stopPart = parts.drop(1).joinToString(";")
        val stopped = findVehicles(stopPart, spec.vehicles).map { it.id }.toSet()
        val f = fold(main)
        val stopsAll = Regex("tat ca (cac )?xe (deu )?(phai )?dung").containsMatchIn(f) || Regex("khong xe nao (duoc )?di").containsMatchIn(f)
        val movesAll = Regex("(hai|ba|bon|cac|tat ca) xe (deu |cung |duoc )?(di |chay )?(cung luc|cung di|duoc di|di cung)").containsMatchIn(f) ||
            Regex("^ca (hai|ba|bon|cac) xe\\.?$").containsMatchIn(f.trim())
        if (stopsAll) return Parsed(emptyList(), movesAll = false, stopsAll = true)
        if (movesAll) return Parsed(listOf(spec.vehicles.filter { it.id !in stopped }.map { it.id }), movesAll = true, stopsAll = false)
        if (Regex("dung lai|phai dung").containsMatchIn(f) && !Regex("duoc di|di truoc|di sau").containsMatchIn(f)) {
            val named = findVehicles(main, spec.vehicles).map { it.id }.toSet()
            if (named.isEmpty()) return null
            val movers = spec.vehicles.filter { it.id !in named && it.id !in stopped }.map { it.id }
            return Parsed(if (movers.isNotEmpty()) listOf(movers) else emptyList(), movesAll = false, stopsAll = false)
        }
        val hits = findVehicles(main, spec.vehicles).filter { it.id !in stopped }
        if (hits.isEmpty()) return null
        val groups = ArrayList<MutableList<String>>()
        hits.forEachIndexed { i, h ->
            if (i == 0) { groups += mutableListOf(h.id); return@forEachIndexed }
            val between = h.text.substring(hits[i - 1].at, h.at)
            val together = Regex("\\b(va|cung|dong thoi)\\b").containsMatchIn(between) && !Regex(",|;| sau | truoc | tiep | roi ").containsMatchIn(between)
            if (together) groups.last() += h.id else groups += mutableListOf(h.id)
        }
        return Parsed(groups, movesAll = false, stopsAll = false)
    }

    private fun axis(dir: String) = if (dir == "N" || dir == "S") "NS" else "EW"

    private fun stopReason(v: JunctionVehicle, spec: JunctionScene): String? {
        if (spec.stopAll) return "phải dừng theo hiệu lệnh tay giơ thẳng đứng"
        spec.lights?.let { lights ->
            when (lights[axis(v.from)]) {
                "red" -> return "đang gặp đèn đỏ"
                "yellow" -> return "gặp đèn vàng (phải dừng trước vạch)"
            }
        }
        if (spec.police != null) return "bị người điều khiển giao thông ra hiệu lệnh dừng"
        return null
    }

    private fun nameOf(spec: JunctionScene, id: String) = spec.vehicles.first { it.id == id }.label

    private val DASH_PREFIX = Regex("^[^—–-]*[—–-]\\s*")
    private val VIOLATION_SUFFIX = Regex("\\s*[—–-]?\\s*vi phạm!?$", RegexOption.IGNORE_CASE)

    private fun stepFor(spec: JunctionScene, id: String): String {
        val gi = spec.order.indexOfFirst { id in it }
        val s = if (gi >= 0) spec.steps?.getOrNull(gi) else null
        return s?.replace(DASH_PREFIX, "")?.replace(VIOLATION_SUFFIX, "") ?: ""
    }

    private fun violationOf(spec: JunctionScene, id: String): String {
        val name = fold(nameOf(spec, id))
        val s = spec.steps?.firstOrNull { fold(it).contains(name) } ?: ""
        return s.replace(VIOLATION_SUFFIX, "").trim()
    }

    private fun correctPlan(spec: JunctionScene, text: String? = null): Plan {
        val viol = spec.violators
        val fallback = when {
            viol.isNotEmpty() -> "Nhận diện đúng: ${viol.joinToString("; ") { violationOf(spec, it).ifBlank { nameOf(spec, it) } }.lowercase()} — đó là hành vi vi phạm."
            spec.stopAll -> "Tất cả các xe dừng lại theo hiệu lệnh — an toàn."
            else -> "Các xe đi đúng thứ tự — qua giao lộ an toàn, không xung đột."
        }
        return Plan(
            correct = true,
            order = spec.order,
            steps = spec.order.mapIndexed { i, g -> spec.steps?.getOrNull(i) ?: g.joinToString(" + ") { nameOf(spec, it) } },
            violators = viol,
            verdict = Consequence("ok", text ?: fallback),
        )
    }

    private fun sameGroups(a: List<List<String>>, b: List<List<String>>) =
        a.size == b.size && a.indices.all { i -> a[i].size == b[i].size && a[i].all { it in b[i] } }

    /** Kịch bản khi chọn đáp án [choice] cho câu sa hình [q]; null nếu câu không có sa hình giao lộ. */
    fun plan(q: Question, choice: Int): Plan? {
        val spec = q.scene as? JunctionScene ?: return null
        val explicit = q.consequences?.getOrNull(choice)
        if (choice == q.answer) return correctPlan(spec, explicit?.text)

        val qText = fold(q.text)
        val chosenText = q.options.getOrNull(choice) ?: ""
        val correctText = q.options.getOrNull(q.answer) ?: ""

        // Câu "xe nào vi phạm": diễn đúng tình huống, kết luận chỉ ra xe vi phạm thật.
        if (qText.contains("vi pham") || (spec.violators.isNotEmpty() && !qText.contains("thu tu"))) {
            val real = spec.violators.map { nameOf(spec, it) }
            val picked = findVehicles(chosenText, spec.vehicles).map { nameOf(spec, it.id) }
            val why = spec.violators.map { violationOf(spec, it) }.filter { it.isNotBlank() }.joinToString("; ")
            val text = explicit?.text ?: if (real.isNotEmpty()) {
                (if (picked.isNotEmpty()) "Bạn chọn \"${chosenText.removeSuffix(".")}\" — chưa đúng. " else "") +
                    "Xe vi phạm là ${real.joinToString(" và ").lowercase()}${if (why.isNotBlank()) ": ${why.lowercase()}" else ""}."
            } else "Không xe nào vi phạm trong tình huống này — đáp án đúng là \"${correctText.removeSuffix(".")}\"."
            return correctPlan(spec).copy(correct = false, verdict = Consequence(explicit?.kind ?: "ticket", text))
        }

        val parsed = parseOption(chosenText, spec)
            ?: return correctPlan(spec).copy(correct = false, verdict = explicit ?: Consequence("danger", "Cách hiểu này chưa đúng. Đáp án đúng: $correctText"))

        if (sameGroups(parsed.groups, spec.order)) return correctPlan(spec, explicit?.text)

        val shouldMove = spec.order.flatten().toSet()
        val done = HashSet<String>()
        val order = ArrayList<List<String>>()
        val steps = ArrayList<String>()

        for (g in parsed.groups) {
            val group = g.filter { it !in done }
            if (group.isEmpty()) continue
            val nextCorrect = spec.order.map { gr -> gr.filter { it !in done } }.firstOrNull { it.isNotEmpty() } ?: emptyList()
            val legal = group.all { it in nextCorrect }
            if (legal) {
                order += group
                steps += if (group.size == nextCorrect.size) {
                    spec.steps?.getOrNull(spec.order.indexOfFirst { gr -> gr.any { it in group } }) ?: group.joinToString(" + ") { nameOf(spec, it) }
                } else group.joinToString(" + ") { nameOf(spec, it) }
                done += group
                continue
            }
            // Có xe đi sai lượt
            val offender = group.first { it !in nextCorrect }
            val offV = spec.vehicles.first { it.id == offender }
            val mustStop = if (offender !in shouldMove) stopReason(offV, spec) else null
            val victim = nextCorrect.firstOrNull()
            val reason = victim?.let { stepFor(spec, it) } ?: ""
            val oName = nameOf(spec, offender)
            val vName = victim?.let { nameOf(spec, it) } ?: ""
            val (kind, text) = if (victim != null) {
                "crash" to (if (mustStop != null)
                    "$oName $mustStop nhưng vẫn đi, cắt ngang ${lowerFirst(vName)} đang được đi${if (reason.isNotBlank()) " (${lowerFirst(reason)})" else ""} — va chạm ngay giữa giao lộ."
                else "$oName đi trước ${lowerFirst(vName)} trong khi ${lowerFirst(vName)} đang có quyền đi trước${if (reason.isNotBlank()) " (${lowerFirst(reason)})" else ""} — va chạm ngay giữa giao lộ.")
            } else {
                "ticket" to (if (mustStop != null) "$oName $mustStop nhưng vẫn đi — vi phạm hiệu lệnh, tín hiệu giao thông." else "$oName đi không đúng thứ tự — vi phạm quy tắc nhường đường.")
            }
            order += if (victim != null) listOf(offender, victim) else listOf(offender)
            steps += if (victim != null) "$oName và ${lowerFirst(vName)} cùng lao vào giao lộ" else "$oName đi sai lượt"
            return Plan(
                correct = false, order = order, steps = steps,
                conflict = victim?.let { Conflict(offender, it, reason) },
                violators = group.filter { it !in nextCorrect },
                verdict = explicit ?: Consequence(kind, text),
            )
        }

        val missing = spec.order.flatten().filter { it !in done }.map { nameOf(spec, it) }
        return Plan(
            correct = false, order = order, steps = steps, violators = emptyList(),
            verdict = explicit ?: Consequence(
                "ok",
                if (missing.isNotEmpty()) "Không gây va chạm, nhưng ${missing.joinToString(", ").lowercase()} không cần dừng lại — đáp án đúng: $correctText"
                else "Thứ tự này chưa đúng — đáp án đúng: $correctText",
            ),
        )
    }
}

File: src/lib/whatif.ts
import type { Consequence, ConsequenceKind, Dir, JunctionScene, JunctionVehicle, Question } from "./types";
import { TOP_LABEL } from "@/components/scene/sprites";

/**
 * Mô phỏng "nếu làm vậy thì sao": từ đáp án người học chọn, suy ra kịch bản chạy sa hình —
 * xe nào đi trước, xe nào bị cắt ngang, xe nào vượt đèn đỏ / không chấp hành CSGT — để
 * JunctionScene diễn lại và cho thấy hậu quả (va chạm, vi phạm) thay vì chỉ chiếu cách đi đúng.
 */

export interface Conflict {
  /** Xe đi sai (đi trước khi chưa được, hoặc đi khi phải dừng). */
  offender: string;
  /** Xe đang có quyền đi mà bị cắt ngang. */
  victim: string;
  /** Lý do xe bị cắt ngang có quyền đi trước (lấy từ `steps` của sa hình). */
  reason: string;
}

export interface WhatIfPlan {
  /** Kịch bản khớp đáp án đúng. */
  correct: boolean;
  /** Các nhóm xe lần lượt di chuyển (nhóm cuối có thể là nhóm xung đột). */
  order: string[][];
  /** Chú thích từng nhóm. */
  steps: string[];
  /** Xung đột ở nhóm cuối cùng (nếu có). */
  conflict?: Conflict;
  /** Xe bị đánh dấu vi phạm trong kịch bản này. */
  violators: string[];
  /** Kết luận hiển thị dưới mô phỏng. */
  verdict: Consequence;
}

/* ---------------- chuẩn hoá & nhận diện nhãn xe ---------------- */

export function fold(s: string) {
  return s
    .normalize("NFD")
    .replace(/[̀-ͯ]/g, "")
    .replace(/đ/g, "d")
    .replace(/Đ/g, "D")
    .toLowerCase();
}

export function vehicleLabel(v: JunctionVehicle) {
  return v.label ?? TOP_LABEL[v.kind];
}

function lower(label: string) {
  return label.charAt(0).toLowerCase() + label.slice(1);
}

/** Vị trí xuất hiện của từng xe trong một đoạn văn bản (theo nhãn), đã tránh nhãn lồng nhau ("Xe con" trong "Xe con A"). */
function findVehicles(text: string, vehicles: JunctionVehicle[]) {
  const hay = fold(text);
  const labels = vehicles
    .map((v) => ({ id: v.id, key: fold(vehicleLabel(v)) }))
    .sort((a, b) => b.key.length - a.key.length);
  const taken: [number, number][] = [];
  const hits: { id: string; at: number }[] = [];
  for (const { id, key } of labels) {
    let from = 0;
    while (from < hay.length) {
      const i = hay.indexOf(key, from);
      if (i < 0) break;
      const overlaps = taken.some(([a, b]) => i < b && i + key.length > a);
      if (!overlaps) {
        taken.push([i, i + key.length]);
        hits.push({ id, at: i });
        break;
      }
      from = i + 1;
    }
  }
  return hits.sort((a, b) => a.at - b.at).map((h) => ({ ...h, text: hay }));
}

/** Phân tích một đáp án thành các nhóm xe đi lần lượt; xe trong phần "…; xe X dừng lại" không đi. */
export function parseOption(option: string, spec: JunctionScene): { groups: string[][]; movesAll: boolean; stopsAll: boolean } | null {
  const [main, ...rest] = option.split(";");
  const stopPart = rest.join(";");
  const stopped = new Set(findVehicles(stopPart, spec.vehicles).map((h) => h.id));
  const f = fold(main);
  const stopsAll = /tat ca (cac )?xe (deu )?(phai )?dung/.test(f) || /khong xe nao (duoc )?di/.test(f);
  const movesAll = /(hai|ba|bon|cac|tat ca) xe (deu |cung |duoc )?(di |chay )?(cung luc|cung di|duoc di|di cung)/.test(f) || /^ca (hai|ba|bon|cac) xe\.?$/.test(f.trim());
  if (stopsAll) return { groups: [], movesAll: false, stopsAll: true };
  if (movesAll) return { groups: [spec.vehicles.filter((v) => !stopped.has(v.id)).map((v) => v.id)], movesAll: true, stopsAll: false };
  // "Chỉ xe con phải dừng lại." → xe được nêu dừng, các xe còn lại đi.
  if (/dung lai|phai dung/.test(f) && !/duoc di|di truoc|di sau/.test(f)) {
    const named = new Set(findVehicles(main, spec.vehicles).map((h) => h.id));
    if (!named.size) return null;
    const movers = spec.vehicles.filter((v) => !named.has(v.id) && !stopped.has(v.id)).map((v) => v.id);
    return { groups: movers.length ? [movers] : [], movesAll: false, stopsAll: false };
  }
  const hits = findVehicles(main, spec.vehicles).filter((h) => !stopped.has(h.id));
  if (!hits.length) return null;
  // Xe nằm trong ngoặc/“dừng lại” ngay trong phần chính cũng không đi.
  const groups: string[][] = [];
  hits.forEach((h, i) => {
    if (i === 0) {
      groups.push([h.id]);
      return;
    }
    const between = h.text.slice(hits[i - 1].at, h.at);
    const together = /\b(va|cung|dong thoi)\b/.test(between) && !/,|;| sau | truoc | tiep | roi /.test(between);
    if (together) groups[groups.length - 1].push(h.id);
    else groups.push([h.id]);
  });
  return { groups, movesAll: false, stopsAll: false };
}

/* ---------------- lý do xe phải dừng ---------------- */

const AXIS: Record<Dir, "NS" | "EW"> = { N: "NS", S: "NS", E: "EW", W: "EW" };

function stopReason(v: JunctionVehicle, spec: JunctionScene): string | null {
  if (spec.stopAll) return "phải dừng theo hiệu lệnh tay giơ thẳng đứng";
  if (spec.lights) {
    const c = spec.lights[AXIS[v.from]];
    if (c === "red") return "đang gặp đèn đỏ";
    if (c === "yellow") return "gặp đèn vàng (phải dừng trước vạch)";
  }
  if (spec.police) return "bị người điều khiển giao thông ra hiệu lệnh dừng";
  return null;
}

function nameOf(spec: JunctionScene, id: string) {
  const v = spec.vehicles.find((x) => x.id === id)!;
  return vehicleLabel(v);
}

function stepFor(spec: JunctionScene, id: string) {
  const gi = spec.order.findIndex((g) => g.includes(id));
  const s = gi >= 0 ? spec.steps?.[gi] : undefined;
  return s ? s.replace(/^[^—–-]*[—–-]\s*/, "").replace(/\s*[—–-]?\s*vi phạm!?$/i, "") : "";
}

/** Mô tả hành vi vi phạm của một xe (từ `steps`, bỏ đuôi "— vi phạm!"). */
function violationOf(spec: JunctionScene, id: string) {
  const name = fold(nameOf(spec, id));
  const s = spec.steps?.find((t) => fold(t).includes(name)) ?? "";
  return s.replace(/\s*[—–-]?\s*vi phạm!?$/i, "").trim();
}

/* ---------------- lập kịch bản ---------------- */

function correctPlan(spec: JunctionScene, text?: string): WhatIfPlan {
  const viol = spec.violators ?? [];
  const fallback = viol.length
    ? `Nhận diện đúng: ${viol.map((id) => violationOf(spec, id) || nameOf(spec, id)).join("; ").toLowerCase()} — đó là hành vi vi phạm.`
    : spec.stopAll
      ? "Tất cả các xe dừng lại theo hiệu lệnh — an toàn."
      : "Các xe đi đúng thứ tự — qua giao lộ an toàn, không xung đột.";
  return {
    correct: true,
    order: spec.order,
    steps: spec.order.map((g, i) => spec.steps?.[i] ?? g.map((id) => nameOf(spec, id)).join(" + ")),
    violators: viol,
    verdict: { kind: "ok", text: text ?? fallback },
  };
}

function sameGroups(a: string[][], b: string[][]) {
  if (a.length !== b.length) return false;
  return a.every((g, i) => g.length === b[i].length && g.every((id) => b[i].includes(id)));
}

/**
 * Kịch bản mô phỏng khi chọn đáp án `choice` cho câu sa hình `q`.
 * Trả về null nếu câu không có sa hình.
 */
export function whatIf(q: Question, choice: number): WhatIfPlan | null {
  const spec = q.scene?.kind === "junction" ? q.scene : null;
  if (!spec) return null;
  const explicit = q.consequences?.[choice] ?? undefined;
  if (choice === q.answer) return correctPlan(spec, explicit?.text);

  const isViolationQ = /vi pham/.test(fold(q.text));
  const chosenText = q.options[choice] ?? "";
  const correctText = q.options[q.answer] ?? "";

  // Câu "xe nào vi phạm": mô phỏng vẫn diễn đúng tình huống, kết luận chỉ ra xe vi phạm thật.
  if (isViolationQ || (spec.violators?.length && !/thu tu/.test(fold(q.text)))) {
    const real = (spec.violators ?? []).map((id) => nameOf(spec, id));
    const picked = findVehicles(chosenText, spec.vehicles).map((h) => nameOf(spec, h.id));
    const why = (spec.violators ?? []).map((id) => violationOf(spec, id)).filter(Boolean).join("; ");
    const text =
      explicit?.text ??
      (real.length
        ? `${picked.length ? `Bạn chọn "${chosenText.replace(/\.$/, "")}" — chưa đúng. ` : ""}Xe vi phạm là ${real.join(" và ").toLowerCase()}${why ? `: ${why.toLowerCase()}` : ""}.`
        : `Không xe nào vi phạm trong tình huống này — đáp án đúng là "${correctText.replace(/\.$/, "")}".`);
    return { ...correctPlan(spec), correct: false, verdict: { kind: explicit?.kind ?? "ticket", text } };
  }

  const parsed = parseOption(chosenText, spec);
  if (!parsed) {
    // Không nhận diện được xe trong đáp án (câu đúng/sai, mô tả chung): chiếu cách đi đúng kèm kết luận.
    return {
      ...correctPlan(spec),
      correct: false,
      verdict: explicit ?? { kind: "danger", text: `Cách hiểu này chưa đúng. Đáp án đúng: ${correctText}` },
    };
  }

  if (sameGroups(parsed.groups, spec.order)) return correctPlan(spec, explicit?.text);

  const shouldMove = new Set(spec.order.flat());
  const done = new Set<string>();
  const order: string[][] = [];
  const steps: string[] = [];

  for (let k = 0; k < parsed.groups.length; k++) {
    const group = parsed.groups[k].filter((id) => !done.has(id));
    if (!group.length) continue;
    // Nhóm đúng tiếp theo: nhóm đầu tiên trong thứ tự đúng còn xe chưa đi.
    const nextCorrect = spec.order.map((g) => g.filter((id) => !done.has(id))).find((g) => g.length) ?? [];
    const legal = group.every((id) => nextCorrect.includes(id));
    if (legal && group.length === nextCorrect.length) {
      order.push(group);
      steps.push(spec.steps?.[spec.order.findIndex((g) => g.some((id) => group.includes(id)))] ?? group.map((id) => nameOf(spec, id)).join(" + "));
      group.forEach((id) => done.add(id));
      continue;
    }
    if (legal) {
      // Đi đúng lượt nhưng thiếu xe cùng lượt — không sai luật; tiếp tục.
      order.push(group);
      steps.push(group.map((id) => nameOf(spec, id)).join(" + "));
      group.forEach((id) => done.add(id));
      continue;
    }
    // Có xe đi sai lượt.
    const offenders = group.filter((id) => !nextCorrect.includes(id));
    const victims = nextCorrect;
    const offender = offenders[0];
    const offV = spec.vehicles.find((v) => v.id === offender)!;
    const mustStop = !shouldMove.has(offender) ? stopReason(offV, spec) : null;
    const victim = victims[0];
    const reason = victim ? stepFor(spec, victim) : "";
    const oName = nameOf(spec, offender);
    const vName = victim ? nameOf(spec, victim) : "";
    let text: string;
    let kind: ConsequenceKind;
    if (victim) {
      kind = "crash";
      text = mustStop
        ? `${oName} ${mustStop} nhưng vẫn đi, cắt ngang ${lower(vName)} đang được đi${reason ? ` (${lower(reason)})` : ""} — va chạm ngay giữa giao lộ.`
        : `${oName} đi trước ${lower(vName)} trong khi ${lower(vName)} đang có quyền đi trước${reason ? ` (${lower(reason)})` : ""} — va chạm ngay giữa giao lộ.`;
    } else {
      kind = "ticket";
      text = mustStop ? `${oName} ${mustStop} nhưng vẫn đi — vi phạm hiệu lệnh, tín hiệu giao thông.` : `${oName} đi không đúng thứ tự — vi phạm quy tắc nhường đường.`;
    }
    order.push(victim ? [offender, victim] : [offender]);
    steps.push(victim ? `${oName} và ${lower(vName)} cùng lao vào giao lộ` : `${oName} đi sai lượt`);
    return {
      correct: false,
      order,
      steps,
      conflict: victim ? { offender, victim, reason } : undefined,
      violators: offenders,
      verdict: explicit ?? { kind, text },
    };
  }

  // Đáp án chỉ khác ở việc bỏ sót xe / dừng xe không cần dừng.
  const missing = spec.order.flat().filter((id) => !done.has(id));
  const stoppedWrong = missing.map((id) => nameOf(spec, id));
  return {
    correct: false,
    order,
    steps,
    violators: [],
    verdict: explicit ?? {
      kind: "ok",
      text: stoppedWrong.length
        ? `Không gây va chạm, nhưng ${stoppedWrong.join(", ").toLowerCase()} không cần dừng lại — đáp án đúng: ${correctText}`
        : `Thứ tự này chưa đúng — đáp án đúng: ${correctText}`,
    },
  };
}
